File: remove-duplicates-from-sorted-list-2.py
# definition for singly linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        prev = dummy        # prev = the last node before the sublist of duplicates

        while head:
            if head.next and head.val == head.next.val:
                while head.next and head.val == head.next.val:      # move till the end of the duplicates sublist
                    head = head.next
                prev.next = head.next       # skip all the duplicates
            else:
                prev = prev.next        # otherwise, move prev
            head = head.next        # move forward
        return dummy.next

File: linkedList/remove_duplicates_from_sorted_list2.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 * 
 * Key Concepts:
 * We need to remove ALL occurrences of duplicated numbers
 * The list is already sorted
 * We need to keep track of the previous non-duplicate node
 * 
 * Algorithm Steps:
 1. Create a dummy head to handle cases where first elements are duplicates
 2. Use two pointers: prev (last confirmed unique node) and current
 3. When duplicates are found:
    Skip all nodes with the same value
    Link prev to the node after duplicates
4. When no duplicates:
    Move prev forward
5. Continue until end of list
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
const deleteDuplicates = (head) => {
  // Handle edge cases
  if (!head || !head.next) return head;

  // Create dummy head to handle cases where first elements are duplicates
  const dummyHead = new ListNode(0);
  dummyHead.next = head;

  let prev = dummyHead;
  let current = head;

  while (current) {
    // Skip all consecutive nodes with same value
    while (current.next && current.val === current.next.val) {
      current = current.next;
    }

    if (prev.next === current) {
      // No duplicates found, move prev
      prev = prev.next;
    } else {
      // Duplicates found, skip all of them
      prev.next = current.next;
    }

    current = current = current.next;
  }

  return dummyHead.next;
};
